datasetIngestor/checkMetadata.go

```diff
@@ -2,111 +2,203 @@ package datasetIngestor
+	DUMMY_TIME = "2300-01-01T11:11:11.000Z"
+	DUMMY_OWNER = "x12345"
 )
 
-const DUMMY_TIME = "2300-01-01T11:11:11.000Z"
-const DUMMY_OWNER = "x12345"
 
-func getHost() string {
-	hostname, err := os.Hostname()
+func CheckMetadata(client *http.Client, APIServer string, metadatafile string, user map[string]string, accessGroups []string) (metaDataMap map[string]interface{}, sourceFolder string, beamlineAccount bool, err error) {
+	metaDataMap, err = readMetadataFromFile(metadatafile)
 	if err != nil {
-		return "unknown"
+			return nil, "", false, err
 	}
 
-	addrs, err := net.LookupIP(hostname)
+	if checkIllegalKeys(metaDataMap) {
+			return nil, "", false, errors.New(ErrIllegalKeys)
+	}
+
+	beamlineAccount, err = checkUserAndOwnerGroup(user, accessGroups, metaDataMap)
 	if err != nil {
-		return hostname
+			return nil, "", false, err
 	}
 
-	for _, addr := range addrs {
-		if ipv4 := addr.To4(); ipv4 != nil {
-			ip, err := ipv4.MarshalText()
-			if err != nil {
-				return hostname
-			}
-			hosts, err := net.LookupAddr(string(ip))
-			if err != nil || len(hosts) == 0 {
-				return hostname
-			}
-			fqdn := hosts[0]
-			return strings.TrimSuffix(fqdn, ".") // return fqdn without trailing dot
-		}
+	err = augmentMissingMetadata(user, metaDataMap, client, APIServer, accessGroups)
+	if err != nil {
+			return nil, "", false, err
 	}
-	return hostname
+
+	err = checkMetadataValidity(client, APIServer, metaDataMap, metaDataMap["type"].(string))
+	if err != nil {
+			return nil, "", false, err
+	}
+
+	sourceFolder, err = getSourceFolder(metaDataMap)
+	if err != nil {
+			return nil, "", false, err
+	}
+
+	return metaDataMap, sourceFolder, beamlineAccount, nil
 }
 
-func CheckMetadata(client *http.Client, APIServer string, metadatafile string, user map[string]string,
-	accessGroups []string) (metaDataMap map[string]interface{}, sourceFolder string, beamlineAccount bool) {
 
-	// read full meta data
+// readMetadataFromFile reads the metadata from the file and unmarshals it into a map.
+func readMetadataFromFile(metadatafile string) (map[string]interface{}, error) {
 	b, err := ioutil.ReadFile(metadatafile) // just pass the file name
 	if err != nil {
 		log.Fatal(err)
 	}
-
-	var metadataObj interface{}
+	// Unmarshal the JSON metadata into an interface{} object.
+	var metadataObj interface{} // Using interface{} allows metadataObj to hold any type of data, since it has no defined methods.
 	err = json.Unmarshal(b, &metadataObj)
 	if err != nil {
 		log.Fatal(err)
 	}
-	// use type assertion to access f's underlying map
-	metaDataMap = metadataObj.(map[string]interface{})
-	beamlineAccount = false
 
+	// Use type assertion to convert the interface{} object to a map[string]interface{}.
+	metaDataMap := metadataObj.(map[string]interface{}) // `.(` is type assertion: a way to extract the underlying value of an interface and check whether it's of a specific type.
+	return metaDataMap, err
+}
+
+func checkIllegalKeys(metadata map[string]interface{}) bool {
```

Comment:
Should document that `true` means illegal. I think I would have expected the opposite.

Reply:
I changed the function name to `containsIllegalKeys` so that true makes sense for illegal.